javalib/src/main/scala/java/util/Formatter.scala

```diff
@@ -79,1587 +91,989 @@ class Formatter(
-        case _ =>
-          throw new UnknownFormatConversionException(
-            String.valueOf(formatToken.getConversionType()))
-      }
+  private def computerizedScientificNotation(
+      x: Double,
+      precision: Int,
+      forceDecimalSep: Boolean): String = {
+    val s1: String = Zone { implicit z =>
+      val buf = z.alloc(2 + precision)
+      vsprintf(buf, toCString(s"%.${precision}e"), toCVarArgList(x))
+      fromCString(buf)
+    }
+
+    // Then make sure the exponent has at least 2 digits for the JDK spec
+    val len = s1.length
+    val s2 =
+      if ('e' != s1.charAt(len - 3)) s1
+      else s1.substring(0, len - 1) + "0" + s1.substring(len - 1)
+
+    // Finally, force the decimal separator, if requested
+    if (!forceDecimalSep || s2.indexOf(".") >= 0) {
+      s2
+    } else {
+      val pos = s2.indexOf("e")
+      s2.substring(0, pos) + "." + s2.substring(pos)
     }
+  }
 
-    def getMinusSign(): Char = minusSign
-
-    def getAddSign(): Char = '+'
-
-    def transform_e(): Unit = {
-      val pattern = new StringBuilder()
-      pattern.append('0')
-      if (formatToken.getPrecision() > 0) {
-        pattern.append('.')
-        val zeros = Array.fill[Char](formatToken.getPrecision())('0')
-        pattern.append(zeros)
-      }
-      pattern.append('E')
-      // Porting note: Harmony appends "+00" here, but it is nonstandard; applyPattern throws IllegalArgumentException on OpenJDK 8
-      pattern.append("00")
-      decimalFormat.applyPattern(pattern.toString())
-      val formattedString = decimalFormat.format(argument)
-      result.append(formattedString.replace("E", "e+").replace("e+-", "e-"))
-
-      if (formatToken.isFlagSet(FormatToken.FLAG_SHARP) && 0 == formatToken
-            .getPrecision()) {
-        val indexOfE = result.indexOf("e")
-        val dot      = decimalFormat.getDecimalFormatSymbols().getDecimalSeparator()
-        result.insert(indexOfE, dot)
-      }
+  private def generalScientificNotation(x: Double,
+                                        precision: Int,
+                                        forceDecimalSep: Boolean): String = {
+    val m = Math.abs(x)
+
+    val p =
+      if (precision == 0) 1
+      else precision
+
+    // between 1e-4 and 10e(p): display as fixed
+    if (m >= 1e-4 && m < Math.pow(10, p)) {
+      /* First approximation of the smallest power of 10 that is >= m.
+       * Due to rounding errors in the event of an imprecise `log10`
+       * function, sig0 could actually be the smallest power of 10
+       * that is > m.
+       */
+      val sig0 = Math.ceil(Math.log10(m)).toInt
+      /* Increment sig0 so that it is always the first power of 10
+       * that is > m.
+       */
+      val sig = if (Math.pow(10, sig0) <= m) sig0 + 1 else sig0
+      decimalNotation(x, Math.max(p - sig, 0), forceDecimalSep)
+    } else if (m == 0.0) { // exact 0 should always be fixed
+      decimalNotation(x, Math.max(precision - 1, 0), forceDecimalSep)
```

Comment:
This was not in the Scala.js code. Was it wrong?

Reply:
Yes, it was not JVM compilent - in JVM `0.0` is always formatted using decimal notation